File: 프로그래머스/lv2/12981. 영어 끝말잇기/영어 끝말잇기.js
function solution(n, words) {
    const answer = [];
    const checkList = {};
    let lastItem = words[0];
    
    for (let i = 0; i < words.length; i++) {
        const turn = i % n;
        const checkItem = words[i];
        
        // 규칙 3번 검사
        if (i !== 0) {
            if (lastItem[lastItem.length - 1] !== checkItem[0]) {
                answer.push(turn + 1, Math.ceil((i+1) / n));
                return answer;
            } else {
                lastItem = checkItem;
            }
        }
        
        // 규칙 5번 검사
        if (checkItem.length === 1) {
            answer.push(turn + 1, Math.ceil((i+1) / n));
            return answer;
        }
        
        // 규칙 4번 검사
        for (const key in checkList) {
            if (checkList[key].includes(checkItem)) {
                answer.push(turn + 1, Math.ceil((i+1) / n))
                return answer;
            }
        }
        
        // 규칙에 걸리지 않았다면 객체에 추가
        if (checkList[turn]) {
            checkList[turn].push(checkItem)
        } else {
            checkList[turn] = [checkItem];
        }       
    }
    
    return [0, 0]

    
    
}

File: 프로그래머스/lv2/12981. 영어 끝말잇기/영어 끝말잇기.py
import math

def solution(n, words):
    words_obj = {}
    
    for i in range(len(words)):
        turn = i % n
        checkItem = words[i]
        
        # 한 글자인 경우
        if len(words[i]) == 1:
            return [turn + 1, math.ceil((i + 1) / n)]

        # 끝말을 맞추지 않은 경우
        if i > 0:
            if words[i][0] != words[i - 1][-1]:
                return [turn + 1, math.ceil((i + 1) / n)]
                
        # 이미 있는 단어를 말한 경우
        if words[i] in words_obj:
            return [turn + 1, math.ceil((i + 1) / n)]
        
        words_obj[words[i]] = 1

    return [0, 0]
